src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java

```diff
@@ -56,6 +55,12 @@
 
     private static final Pattern WHITESPACES_PATTERN = Pattern.compile("\\p{Space}+");
 
+    private static final Pattern JMETER_VARIABLE_REFERENCE_PATTERN = Pattern
+            .compile("(:\\s*)\\$\\{([^\\$\\{\\}]+)\\}([\\s,\\}])");
```

Comment:
The part `[^\\$\\{\\}]+` seems not correct to me. I believe, that you wanted to simulate a *negative lookahead* here.
You probably want to match a complete variable reference like `${varname}` and not two or more like in `${var1} ${var2}`.
What about more complex setups like `${function(${var1})}`?

Reply:
Hi @FSchumacher ,

I'm very sorry for late response.
My intent there is to find a string, (a) starting with "${", followed by (b) any char(s) excluding '$', '{' and '}', followed by (c) "}".  The negative part is only for (b): var name shouldn't include the special chars. I believe it should work even if it looks a bit complex.

If we want to create Pattern objects every time instead of having just one generic Pattern, I'm not so sure how it could improve the readability because the varname should probably be dynamically concatenated at runtime to build a regex pattern, which might be less readable in the end, IMHO.

Regarding the complex setup example with function, is it possible to use that complex variable expression in JMeter? (If so, perhaps we might need to replace the regex by commons-text's StrSubstitutor instead.

Regards,

Woonsan

Reply:
No problem :)

About the regex. In regex a `[` starts a character group, which means, that the order of the elements inside the brackets is irrelevant. 
If you look at the following examples:
```
abc: ${__log("Hello\, ${__machineIP()}")}
abc: ${__log("Some $ and some { or even }")}
abc: ${someFunction(abc)}${anotherFunction}
abc: ${func2(a,b)}
abc: ${func1(a)}
abc: ${func0()}
abc: ${more(abc,)}
abc: Nothing to be found here
```
 `(:\s*)\$\{([^\$\{\}]+)\}([\s,\}])` would match the part after `abc` in lines
```
abc: ${func2(a,b)}
abc: ${func1(a)}
abc: ${func0()}
abc: ${more(abc,)}
```
and a regex with negative lookahead like `:\s*\$\{(?!.*\$\{).*\)}` would match 
```
abc: ${__log("Some $ and some { or even }")}
abc: ${func2(a,b)}
abc: ${func1(a)}
abc: ${func0()}
abc: ${more(abc,)}
```
And those examples don't include a simple variable like `${var}`, which would yield wrong results with the lookahead variant.

About the complex function setup. Considering the usage of the `__log` function, I think it is not that complex. The following expression is a valid one: `${__log("Hello\, ${__machineIP()}")}`.
Another example is given in the help for [`__eval`](https://jmeter.apache.org/usermanual/functions.html#__eval)

Given this, I don't think, that a simple regex approach would help all cases. It might be enough to help most users, though.

Reply:
I see. Thanks a lot for the explanations, @FSchumacher !
Perhaps shall we just avoid Jackson object serialization/deserialization, without having to validate or reformat the json variables at all? It would make it work at least...

Reply:
I would favour dropping the step, as I can't see a good way to keep the functionality *and* make it work in all cases.

Reply:
Hi @FSchumacher ,

I've updated my PR, not to parse/serialize variables, and now it seems working fine with any var expressions in variables or even in query.

Please have a look.

Thanks in advance,

Woonsan